tcljava/doc/TclList.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclList.3,v 1.1 1998/10/14 21:09:09 cvsadmin Exp $
'\" 
.so man.macros
.TH TclList 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclList -- This class is used to create, modify and read from Tcl list
objects. 

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static TclObject \fBnewInstance\fR()
.sp
static void \fBappend\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, TclObject \fIelem\fR)
throws TclException
.sp
static int \fBgetLength\fR(Interp \fIinterp\fR, TclObject \fItobj\fR)
throws TclException
.sp
static TclObject[ ] \fBgetElements\fR(Interp \fIinterp\fR, TclObject \fItobj\fR)
throws TclException
.sp
static TclObject \fBindex\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, int \fIindex\fR)
throws TclException
.sp
static void \fBreplace\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, int \fIindex\fR, int \fIcount\fR, TclObject \fIelts[]\fR, int \fIfrom\fR, int \fIto\fR)

.SH ARGUMENTS
.AS TclObject "elts[]" .
.AP Interp interp
If an error occurs while converting an object to be a list object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
Points to the list object to be manipulated.
If \fItobj\fR does not already point to a list object,
an attempt will be made to convert it to one.
.AP TclObject elem
Points to a list object containing elements to be appended onto \fItobj\fR.
.AP TclObject elts[]
An array of TclObjects to reaplace those in the list object \fItobj\fR.
.AP int index
The index of the elemtent of \fItobj\fR in question.
.AP int count
The number of elemtents of the \fItobj\fR to replace.
.AP int from
The index of the first replacement element in \fIelts\fR.
.AP int to
The index of the last replacement element in \fIelts\fR.
.BE

.SH DESCRIPTION
.PP
The TclList class is used to create, modify and read from Tcl list
objects. All of the methods defined below attempt to change the
internal representation of \fItobj\fR parameters to TclList. If such a
conversion fails (e.g., the string value of \fItobj\fR is not a proper Tcl
list), a TclException is thrown, and an appropriate error message
is left in \fIinterp\fR's result if \fIinterp\fR is not null.
.PP
Tcl list objects have an internal representation that supports
the efficient indexing and appending.
The procedures described in this man page are used to
create, modify, index, and append to Tcl list objects from Java code.
In all cases where \fItobj\fR is an argument, if \fItobj\fR is not a
TclList, then an attempt is made to convert it to a TclList.  If
the attempt fails, a TclException is thrown.
.TP
\fBnewInstance\fR
The \fBnewInstance\fR method creates and returns a new empty TclList.
.TP
\fBappend\fR
The \fBappend\fR method adds one object, \fIelem\fR, to the end of the list
object referenced by \fItobj\fR.  After appending the element,
\fBappend\fR increments the element's reference count since \fItobj\fR
now also refers to it.
.TP
\fBgetLength\fR
The \fBgetLength\fR method returns the number of elements in \fItobj\fR.
.TP
\fBgetElements\fR
The \fBgetElements\fR method converts the TclList \fItobj\fR to an array of
TclObjects.  The objects referenced by the returned array should be
treated as readonly, and their ref counts are not incremented.
.TP
\fBindex\fR
The \fBindex\fR method returns a pointer to the \fIindex\fRth element of
\fItobj\fR.  If \fIindex\fR is negative or greater than or equal to
the number of elements in the list, null is returned.
.TP
\fBreplace\fR
The \fBreplace\fR mehtod replaces zero or more elements of the list referenced by
\fItobj\fR with the objects in \fIelts\fR ranging from \fIfrom\fR to
\fIto\fR.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_NewListObj, Tcl_ListObjAppendList, Tcl_ListObjLength,
Tcl_ListObjGetElements, Tcl_ListObjIndex, Tcl_ListObjReplace

.SH "SEE ALSO"
InternalRep, TclObject, TclIndex, TclDouble, TclString, TclInteger, TclBoolean, ReflectObject

.SH KEYWORDS
append, index, insert, internal representation, length, list, list object, list type, object, object type, replace


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tclblend/tcl/lang/TclList.java
/* 
 * TclList.java --
 *
 *	This file contains the native implementation of Tcl lists.
 *
 * Copyright (c) 1997 by Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * RCS: @(#) $Id: TclList.java,v 1.1 1998/10/14 21:09:10 cvsadmin Exp $
 */

package tcl.lang;

import java.util.*;

/*
 * This class implements the list object type in Tcl.  It is based on the
 * native list implementation, so it extends CObject to inherit the
 * shadowing behavior.
 */

public class TclList extends CObject {

/*
 *----------------------------------------------------------------------
 *
 * TclList --
 *
 *	This constructor creates a new Tcl_Obj.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

TclList()
{
    super();
}

/*
 *----------------------------------------------------------------------
 *
 * TclList --
 *
 *	Construct a new TclList from the given Tcl_Obj reference.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Increments the reference count of the Tcl_Obj.
 *
 *----------------------------------------------------------------------
 */

TclList(
    long objPtr)		// Pointer to Tcl_Obj from C.
{
    super(objPtr);
}

/*
 *----------------------------------------------------------------------
 *
 * newInstance --
 *
 *	Construct a new TclObject containing an empty TclList.
 *
 * Results:
 *	Returns a new TclObject.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

public static TclObject
newInstance()
{
    return new TclObject(new TclList());
}

/*
 *----------------------------------------------------------------------
 *
 * setListFromAny --
 *
 *	Called to convert an object's internal rep to a list.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

static void
setListFromAny(
    Interp interp,		// Interp to report errors in, or null.
    TclObject tobj)		// Object to convert.
throws
    TclException		// Throws an exception of the object isn't a
				// valid Tcl list.
{
    InternalRep rep = tobj.getInternalRep();

    /*
     * If the object is already a Tcl_Obj reference, then we only need to check
     * to see if it is a valid list and copy the reference into a new TclList.
     * Otherwise we need to create a new Tcl_Obj to hold the list.
     */

    if (!(rep instanceof TclList)) {
	TclList tlist;
	long interpPtr = (interp == null) ? 0 : interp.interpPtr;

	if (rep instanceof CObject) {
	    CObject cobj = (CObject) rep;
	    listLength(interpPtr, cobj.objPtr);
	    tlist = new TclList(cobj.objPtr);
	} else {
	    tlist = new TclList(splitList(interpPtr, tobj.toString()));
	}
	tobj.setInternalRep(tlist);
    }
}

/*
 *----------------------------------------------------------------------
 *
 * append --
 *
 *	Appends a list element to a TclObject.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Modifies the underlying Tcl_Obj list.
 *
 *----------------------------------------------------------------------
 */

public static final void
append(
    Interp interp,		// Current interpreter.
    TclObject tobj,		// The TclObject to append an element to.
    TclObject elemObj)		// The element to append to the object.
throws
    TclException		// If tobj cannot be converted into a list.
{
    setListFromAny(interp, tobj);
    tobj.invalidateStringRep();

    TclList tlist = (TclList)tobj.getInternalRep();

    /*
     * We have to update the objPtr after the append, because the append may
     * duplicate the Tcl_Obj if it was shared at the C level.  Because the
     * TclList acts as an extra reference on the Tcl_Obj, we are guaranteed
     * that there will be no circular references when we are done.  After
     * append returns, the refcount on the Tcl_Obj will be 1.
     */

    tlist.objPtr = append(tlist.objPtr, elemObj);
}

/*
 *----------------------------------------------------------------------
 *
 * getLength --
 *
 *	Queries the length of the list. If tobj is not a list object,
 *	an attempt will be made to convert it to a list.
 *
 * Results:
 *	The length of the list.
 *
 * Side effects:
 *	Will attempt to convert the object into a TclList.
 *
 *----------------------------------------------------------------------
 */

public static final int
getLength(
    Interp interp,		// Current interpreter.
    TclObject tobj)		// The TclObject to append an element to.
throws
    TclException		// If tobj is not a valid list.
{
    long interpPtr = (interp == null) ? 0 : interp.interpPtr;
    setListFromAny(interp, tobj);

    TclList tlist = (TclList)tobj.getInternalRep();
    return listLength(interpPtr, tlist.objPtr);
}

/*
 *----------------------------------------------------------------------
 *
 * getElements --
 *
 *	Retrieve all of the elements of a list.  The objects referenced
 *	by the returned array should be treated as readonly and their
 *	ref counts are _not_ incremented; the caller must do that if
 *	it holds on to a reference.
 *
 * Results:
 *	Returns a TclObject array of the elements in the list object.
 *
 * Side effects:
 *	Attempts to convert the object to a list and may throw an
 *	exception if the object isn't a list.  
 *
 *----------------------------------------------------------------------
 */

public static TclObject[]
getElements(
    Interp interp,		// Current interpreter.
    TclObject tobj)		// The TclObject to get the elements of.
throws
    TclException		// If tobj is not a valid list.
{
    setListFromAny(interp, tobj);
    TclList tlist = (TclList)tobj.getInternalRep();
    return getElements(tlist.objPtr);
}

/*
 *----------------------------------------------------------------------
 *
 * index --
 *
 *	Retrieve the index'th object from the list.The first element has
 *	index 0. If index is negative or greater than or equal to the
 *	number of elements in the list, a null is returned.
 *
 * Results:
 *	Returns the index'th TclObject in the list.
 *
 * Side effects:
 *	Attempts to convert the object to a list and may throw an
 *	exception if the object isn't a list.  
 *
 *----------------------------------------------------------------------
 */

public static final TclObject
index(
    Interp interp,		// Current interpreter.
    TclObject tobj,		// The TclObject to get the element from.
    int index)			// The index of the requested element.
throws
    TclException		// If tobj is not a valid list.
{
    setListFromAny(interp, tobj);

    TclList tlist = (TclList)tobj.getInternalRep();
    return index(tlist.objPtr, index);
}

/*
 *----------------------------------------------------------------------
 *
 * replace --
 *
 *	This procedure replaces zero or more elements of the list
 *	referenced by tobj with the objects from an TclObject array.
 *	If tobj is not a list object, an attempt will
 *	be made to convert it to a list.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Updates the Tcl_Obj list.
 *
 *----------------------------------------------------------------------
 */

public static final void
replace(
    Interp interp,		// Current interpreter.
    TclObject tobj,		// The TclObject to get the element from.
    int index,			// The starting index of the replace
				// operation. <=0 means the beginning of the
				// list. >= TclList.getLength(tobj) means the
				// end of the list.
    int count,			// the number of elements to delete from the
				// list. <=0 means no elements should be
				// deleted and the operation is equivalent to
				// an insertion operation.
    TclObject elements[],	// The element(s) to insert.
    int from,			// Insert starting at elements[from]
    int to)			// Insert up to elements[to] (inclusive)
throws
    TclException		// If tobj is not a valid list.
{
    setListFromAny(interp, tobj);
    tobj.invalidateStringRep();
    TclList tlist = (TclList)tobj.getInternalRep();
    tlist.objPtr = replace(tlist.objPtr, index, count, elements, from, to);
}

/*
 * Native method declarations.
 */

private static final native long append(long objPtr, TclObject element);
private static final native TclObject[] getElements(long objPtr);
private static final native TclObject index(long objPtr, int index);
private static final native int listLength(long interp, long objPtr)
	throws TclException;
private static final native long replace(long objPtr, int index, int count,
	TclObject elements[], int from, int to);
private static final native long splitList(long interp, String s);

} // end TclList

